### DISCUSSION TITLE
clients management

### DISCUSSION THREAD

**Discussion by Gianluca-Casagrande-Stiga** [moscajs/aedes#742] [Q&A] [answered]
Hi, I would like to build a broker clients management tool. My targets are:

- get clients list with their subscriptions 
- force a client disconnection

I know that Ades object has "clients" property with every clients info. But in a Redis cluster scenario, this "clients" property is related only to that broker instance, isn't it? Is there a way to get every cluster clients info? Maybe from Redis?

thanks

**Comment by robertsLando** [comment, marked as answer]
You could build a generic one by just using aedes-persistence methods. For example in order to get all subscriptions you could call [subscriptionsByTopic](https://github.com/moscajs/aedes-persistence/blob/main/persistence.js#L165) method with `'#'`.

In order to force a client disconnection you should use the subscription client id to get the client instance from aedes broker and call the `close` method.

Let me know if you build it we could add it to moscajs org if it is an OSS project :)

**Reply by robertsLando** [reply]
In a cluster env if you want to disconnect a specific clent but you dunno which broker instance have it you could use broker to publish on a specific `$SYS` topic like `$SYS/clients/disconnect` with client id as payload, on each instance simply subscribe broker to this topic and when a message is received do your disconnect logic by checking clients instances on that instance. Giving that `$SYS` topics are reserved for internal usage clients will never receive those messages

**Comment by Gianluca-Casagrande-Stiga** [comment]
Thank you Daniel for your hints.
Unfortunately I wasn't able to use subscriptionsByTopic# method on a Redis cluster solution. I've resolved my case with a broker subscription extension combined with a special client (ClientManager). This is my solution explanation:

Each broker of the cluster is subscribed to "$SYS/clients/list" and "$SYS/clients/disconnect" topics.
When a broker receives a message on "$SYS/clients/list", it publish every clients info on topic "$SYS/clients/list/ack".
When a broker receives a message on "$SYS/clients/disconnect", it try to close the client with the id contained in the payload; if the client is successfully closed, it returns an ack on "$SYS/clients/disconnect/ack".

ClientManager is an mqtt client with special auth that let it pub/sub on $SYS topics.
It can get the broker cluster client list publishing a "$SYS/clients/list" request and waiting the "$SYS/clients/list/ack" responses.
It can publish a "$SYS/clients/disconnect" request, waiting a "$SYS/clients/disconnect/ack" response.

**Reply by robertsLando** [reply]
Thanks for sharing :)
